Common++/header/ObjectPool.h

```diff
@@ -0,0 +1,157 @@
+#pragma once
+
+#include <stack>
+#include <mutex>
+#include <memory>
+#include <type_traits>
+
+namespace pcpp
+{
+	/**
+	 * @brief A generic object pool implementation.
+	 *
+	 * This class provides a generic object pool that can be used to efficiently manage and reuse objects of any type.
+	 * Objects can be acquired from the pool using the `acquireObject` method, and released back to the pool using the
+	 * `releaseObject` method. If the pool is empty when acquiring an object, a new object will be created. If the pool
+	 * is full when releasing an object, the object will be deleted.
+	 *
+	 * @tparam T The type of objects managed by the pool. Must be default constructable.
+	 */
+	template <class T, typename std::enable_if<std::is_default_constructible<T>::value, bool>::type = true>
+	class ObjectPool
+	{
+	public:
+		constexpr static std::size_t DEFAULT_POOL_SIZE = 100;
+		constexpr static std::size_t INFINITE_POOL_SIZE = 0;
+
+		/**
+		 * A constructor for this class that creates a pool of objects
+		 * @param[in] maxPoolSize The maximum number of objects in the pool
+		 * @param[in] preallocate The number of objects to preallocate in the pool
+		 */
+		explicit ObjectPool(std::size_t maxPoolSize = DEFAULT_POOL_SIZE, std::size_t preallocate = 0)
+		    : m_maxPoolSize(maxPoolSize)
+		{
+			this->preallocate(preallocate);
+		}
+
+		// These don't strictly need to be deleted, but don't need to be implemented for now either.
+		ObjectPool(const ObjectPool&) = delete;
+		ObjectPool(ObjectPool&&) = delete;
+		ObjectPool& operator=(const ObjectPool&) = delete;
+		ObjectPool& operator=(ObjectPool&&) = delete;
+
+		/**
+		 * A destructor for this class that deletes all objects in the pool
+		 */
+		~ObjectPool()
+		{
+			clear();
+		}
+
+		/**
+		 * @brief Acquires a unique pointer to an object from the pool.
+		 *
+		 * This method acquires a unique pointer to an object from the pool.
+		 * If the pool is empty, a new object will be created.
+		 *
+		 * @return A unique pointer to an object from the pool.
+		 */
+		std::unique_ptr<T> acquireObject()
+		{
+			return std::unique_ptr<T>(acquireObjectRaw());
+		}
+
+		/**
+		 * @brief Acquires a raw pointer to an object from the pool.
+		 *
+		 * This method acquires a raw pointer to an object from the pool.
+		 * If the pool is empty, a new object will be created.
+		 *
+		 * @return A raw pointer to an object from the pool.
+		 */
+		T* acquireObjectRaw()
+		{
+			std::unique_lock<std::mutex> lock(m_mutex);
+
+			if (m_pool.empty())
+			{
+				// We don't need the lock anymore, so release it.
+				lock.unlock();
```

Comment:
The `new T()` operation will be under mutex lock in that case, which we don't want as it might be expensive in relation to the other operations in the function. The critical section ends when the pool determines that the context creation will be delegated to the heap allocator.

Reply:
while I agree with this statement, I'm not sure the performance gain is significant between the 2 options, and IMO, the simpler / more readable option is not to explicitly call `unlock()`. Having said that, I don't have a strong opinion about it, so if you insist this is better I'm ok with that

Reply:
Benchmarked the pool `acquire`/`release` perf and the new/delete perf (combined as I assume for each newly created context above the max, it will have to be deleted too).

All benchmarks are on MSVC x64 Release.
- `new`/`delete` (warmup 5000, count 30000)
  - Total time: 9390600ns
  - Avg time: 313ns
- Pool `acquireObjectRaw`/`releaseObjectRaw` (warmup 5000, count 30000) - Single thread, only pooled items used.
  - Total time: 1191600ns
  - Avg time: 39ns

So essentially, if the `new T()` is done under mutex lock, that essentially blocks all acquire/release operations in the pool for 8x times the length of one non-heap branch acquire/release cycle.

Benchmark sample code
```
#include <iostream>
#include <chrono>
#include "Logger.h"

template <class T> void benchmark(int warmup, int count, T fn)
{
	for (size_t i = 0; i < warmup; i++)
	{
		fn();
	}

	auto clk = std::chrono::high_resolution_clock();
	auto start = clk.now();

	for (size_t i = 0; i < count; i++)
	{
		fn();
	}

	auto end = clk.now();
	std::cout << "Total time: " << (end - start).count() << "ns" << std::endl;
	std::cout << "Avg time: " << (end - start).count() / count << "ns" << std::endl;
}

void benchmarkNewDelete()
{
	constexpr auto W = 5000;
	constexpr auto N = 30000;

	std::cout << "Benchmarking new/delete\n";
	benchmark(W, N, []() {
		pcpp::internal::LogContext* ctx = new pcpp::internal::LogContext();
		delete ctx;
	});
}

void benchmarkPool()
{
	constexpr auto W = 5000;
	constexpr auto N = 30000;

	pcpp::ObjectPool<pcpp::internal::LogContext> pool(5, 2);

	std::cout << "Benchmarking pool\n";
	benchmark(W, N, [&pool]() {
		auto* ctx = pool.acquireObjectRaw();
		pool.releaseObjectRaw(ctx);
	});
}

int main()
{
	// std::ios_base::sync_with_stdio(false);
	using namespace pcpp;

	benchmarkNewDelete();
	benchmarkPool();

	return 0;
}
```